saleor/graphql/product/mutations/products.py

```diff
@@ -78,9 +80,21 @@ class CategoryInput(graphene.InputObjectType):
     seo = SeoInput(description="Search engine optimization fields.")
     background_image = Upload(description="Background image file.")
     background_image_alt = graphene.String(description="Alt text for a product media.")
+    metadata = NonNullList(
+        MetadataInput,
+        description=("Fields required to update the category metadata." + ADDED_IN_38),
+        required=False,
+    )
+    private_metadata = NonNullList(
+        MetadataInput,
+        description=(
+            "Fields required to update the category private metadata." + ADDED_IN_38
+        ),
+        required=False,
+    )
 
 
-class CategoryCreate(ModelMutation):
+class CategoryCreate(MutationWithMetadataMixin, ModelMutation):
```

Comment:
I am not convinced about the current approach. Some metadata logic is placed in `MutationWithMetadataMixin` and some logic is placed `ModelMutation`.  It is hard to get full context jumping between mixin and model mutation. 
Can't we extend `ModelMutation` to have full metadata support?
The support should cover, validating input, validating permission for given instance, saving in the instance etc. 
We could control it by setting the flag in Meta input. Something like:
```
class CategoryCreate(ModelMutation):
    class Arguments:
         ...
         <some input with metadata>
    class Meta:
        ...
        support_meta_field=True,
        support_private_meta_field=True
```
Based on provided flags (i didn't think a lot about the naming of the fields, just use the first one to show the proposed solution), ModelMutation can control if should handle metadata fields from the input or not. Thanks to this, as a developer, one thing that you need to remember during adding new meta fields is to provide correct input fields and setting the flags to correct value.
I think that we could provide similar flow to BaseMutation but more generic (Probably we could add it only to BaseMutation, and use it in ModelMutation)

@SzymJ @maarcingebala  What do you think?

Reply:
cc. @fowczarek as he's the author of the current metadata solution :)

Reply:
IMO solution proposed by @korycins is better.

Reply:
Done, I moved all methods to `BaseMutation`